# Coding Exercises

Now, you will have the opportunity to analyze your own dataset. <br>
__Follow the instructions__ and insert your code! You are either requested to 
- Complete the Code and __Fill in the gaps__. Gaps are marked with "__---__" and are __placeholders__ for your code fragment 
- Write Code completely __on your own__ 

In some exercises, you will find questions that can only be answered, if your code is correct and returns the right output! The correct answer is provided below your coding cell. There you can check whether your code is correct.

### Have Fun!

--------------------------------------------------------------------------------------------------------------

## Time Series Data in Pandas: Introduction


### Exercise 1: Intro to Time Series Data & DatetimeIndex

In [1]:
#run the cell!
import pandas as pd

1. __Import__ the csv-file "__aapl_fb.csv__" containing Stock Price Data for Apple and Facebook for the years 2015-2018! Convert the datatype in the column __Date__ to __datetime__ and set the column __Date__ as the __Index__ of the DataFrame! <br>
__Save__ the DataFrame in the variable __stocks__! __Fill in the gaps!__ 

In [5]:
stocks = pd.read_csv("aapl_fb.csv",parse_dates=['Date'],index_col='Date')

How many Timestamps does __stocks__ contain? What was the __maximum stock price__ for Facebook in that Period?

In [6]:
# run the cell!
stocks.head()

AAPL         FB
Date                             
2015-01-02  109.330002  78.449997
2015-01-05  106.250000  77.190002
2015-01-06  106.260002  76.150002
2015-01-07  107.750000  76.150002
2015-01-08  111.889999  78.180000

In [7]:
# run the cell!
stocks.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 1005 entries, 2015-01-02 to 2018-12-28
Data columns (total 2 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   AAPL    1005 non-null   float64
 1   FB      1005 non-null   float64
dtypes: float64(2)
memory usage: 23.6 KB


In [8]:
# run the cell!
stocks.describe()

AAPL           FB
count  1005.000000  1005.000000
mean    140.988338   133.415502
std      35.082112    35.846003
min      90.339996    74.050003
25%     112.120003   105.019997
50%     129.360001   129.070007
75%     168.630005   167.410004
max     232.070007   217.500000

There are in total ... 1005 Timestamps with stock price data. Facebooks´s maximum stock price was ... 217.50!

2. __Select__ the stock prices on __May 13, 2016__! What was the Price for __Apple__?

In [9]:
stocks.loc["2016-05-13"]

AAPL     90.519997
FB      119.809998
Name: 2016-05-13 00:00:00, dtype: float64

The Apple Stock Price on that day was ... 90.52

3. __Select__ all stock prices in __September 2017__ and calculate the __mean__ for both stocks! What was the __mean stock price__ for __Apple__ in __September 2017__?

In [11]:
stocks.loc["2017-09"].mean()

AAPL    157.6085
FB      170.5910
dtype: float64

Apple´s monthly mean price for September 2017 is ... 157.61! 

4. __Create a Timestamp__ for August 10, 2019! (convert __2 alternative date strings__ of your choice) 

In [12]:
# Alt 1
pd.to_datetime("2019-08-10")

Timestamp('2019-08-10 00:00:00')

In [13]:
# Alt 2
pd.to_datetime("10 Aug 2019")

Timestamp('2019-08-10 00:00:00')

5. __Change Frequency__ of stocks to __monthly__ ("Downsampling") and display the __last__ traded stock price per month! (Hint: The Index should show the last calender day of each month) <br>
What is the __last__ traded price for __Facebook__ in __March 2015__?

In [16]:
stocks.resample("ME").last() 

AAPL          FB
Date                              
2015-01-31  117.160004   75.910004
2015-02-28  128.460007   78.970001
2015-03-31  124.430000   82.220001
2015-04-30  125.150002   78.769997
2015-05-31  130.279999   79.190002
2015-06-30  125.430000   85.769997
2015-07-31  121.300003   94.010002
2015-08-31  112.760002   89.430000
2015-09-30  110.300003   89.900002
2015-10-31  119.500000  101.970001
2015-11-30  118.300003  104.239998
2015-12-31  105.260002  104.660004
2016-01-31   97.339996  112.209999
2016-02-29   96.690002  106.919998
2016-03-31  108.989998  114.099998
2016-04-30   93.739998  117.580002
2016-05-31   99.860001  118.809998
2016-06-30   95.599998  114.279999
2016-07-31  104.209999  123.940002
2016-08-31  106.099998  126.120003
2016-09-30  113.050003  128.270004
2016-10-31  113.540001  130.990005
2016-11-30  110.519997  118.419998
2016-12-31  115.820000  115.050003
2017-01-31  121.349998  130.320007
2017-02-28  136.990005  135.539993
2017-03-31  143.660004  142.050003
2017-04-30  143.649994  150.250000
2017-05-31  152.759995  151.460007
2017-06-30  144.020004  150.979996
2017-07-31  148.729996  169.250000
2017-08-31  164.000000  171.970001
2017-09-30  154.119995  170.869995
2017-10-31  169.039993  180.059998
2017-11-30  171.850006  177.179993
2017-12-31  169.229996  176.460007
2018-01-31  167.429993  186.889999
2018-02-28  178.119995  178.320007
2018-03-31  167.779999  159.789993
2018-04-30  165.259995  172.000000
2018-05-31  186.869995  191.779999
2018-06-30  185.110001  194.320007
2018-07-31  190.289993  172.580002
2018-08-31  227.630005  175.729996
2018-09-30  225.740005  164.460007
2018-10-31  218.860001  151.789993
2018-11-30  178.580002  140.610001
2018-12-31  156.229996  133.199997

The last traded price is ... 82.22!

6. Calculate __monthly average/mean stock prices__ and save the new DataFrame in the variable __stocks_m__!

In [17]:
stocks_m = stocks.resample("ME").mean()

What was the average stock price for __Facebook__ in __April 2015__?

In [23]:
# run the cell!
stocks.loc["April 2015"].mean()

AAPL    127.291428
FB       82.091428
dtype: float64

The average price in april is ... 82.09!

7. __Downsample__ stocks to __weekly__ frequency, showing prices on __Fridays__!
What was __Facebook´s__ stock price on the __fourth Friday__ in __January 2015__?

In [26]:
list(stocks.resample('W-Fri'))

[(Timestamp('2015-01-02 00:00:00'),
                    AAPL         FB
  Date                             
  2015-01-02  109.330002  78.449997),
 (Timestamp('2015-01-09 00:00:00'),
                    AAPL         FB
  Date                             
  2015-01-05  106.250000  77.190002
  2015-01-06  106.260002  76.150002
  2015-01-07  107.750000  76.150002
  2015-01-08  111.889999  78.180000
  2015-01-09  112.010002  77.739998),
 (Timestamp('2015-01-16 00:00:00'),
                    AAPL         FB
  Date                             
  2015-01-12  109.250000  76.720001
  2015-01-13  110.220001  76.449997
  2015-01-14  109.800003  76.279999
  2015-01-15  106.820000  74.050003
  2015-01-16  105.989998  75.180000),
 (Timestamp('2015-01-23 00:00:00'),
                    AAPL         FB
  Date                             
  2015-01-20  108.720001  76.239998
  2015-01-21  109.550003  76.739998
  2015-01-22  112.400002  77.650002
  2015-01-23  112.980003  77.830002),
 (Timestamp('2015-01

Facebook´s stock price on the fourth Friday in January 2015 was ... 77.83!

# Solutions (Stop here if you want to code on your own!)

# ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

## Time Series Data in Pandas: Introduction


### Exercise 1: Intro to Time Series Data & DatetimeIndex

In [ ]:
#run the cell!
import pandas as pd

1. __Import__ the csv-file "__aapl_fb.csv__" containing Stock Price Data for Apple and Facebook for the years 2015-2018! Convert the datatype in the column __Date__ to __datetime__ and set the column __Date__ as the __Index__ of the DataFrame! <br>
__Save__ the DataFrame in the variable __stocks__! __Fill in the gaps!__ 

In [ ]:
stocks = pd.read_csv("aapl_fb.csv", parse_dates= ["Date"], index_col = "Date")

How many Timestamps does __stocks__ contain? What was the __maximum stock price__ for Facebook in that Period?

In [ ]:
# run the cell!
stocks.head()

In [ ]:
# run the cell!
stocks.info()

In [ ]:
# run the cell!
stocks.describe()

There are in total ... 1005 Timestamps with stock price data. Facebooks´s maximum stock price was ... 217.50!

2. __Select__ the stock prices on __May 13, 2016__! What was the Price for __Apple__?

In [ ]:
stocks.loc["2016-05-13"]

The Apple Stock Price on that day was ... 90.52

3. __Select__ all stock prices in __September 2017__ and calculate the __mean__ for both stocks! What was the __mean stock price__ for __Apple__ in __September 2017__?

In [ ]:
stocks.loc["2017-09"].mean()

Apple´s monthly mean price for September 2017 is ... 157.61! 

4. __Create a Timestamp__ for August 10, 2019! (convert __2 alternative date strings__ of your choice) 

In [ ]:
# Alt 1
pd.to_datetime("2019-08-10")

In [ ]:
# Alt 2
pd.to_datetime("10 Aug 2019")

5. __Change Frequency__ of stocks to __monthly__ ("Downsampling") and display the __last__ traded stock price per month! (Hint: The Index should show the last calender day of each month) <br>
What is the __last__ traded price for __Facebook__ in __March 2015__?

In [ ]:
stocks.resample("M").last() # old

In [ ]:
stocks.resample("ME").last() # NEW (Pandas 3.x)

The last traded price is ... 82.22!

6. Calculate __monthly average/mean stock prices__ and save the new DataFrame in the variable __stocks_m__!

In [ ]:
stocks_m  = stocks.resample("ME").mean() # NEW

What was the average stock price for __Facebook__ in __April 2015__?

In [ ]:
# run the cell!
stocks_m.head()

The average price in april is ... 82.09!

7. __Downsample__ stocks to __weekly__ frequency, showing prices on __Fridays__!
What was __Facebook´s__ stock price on the __fourth Friday__ in __January 2015__?

In [ ]:
stocks.resample("W-Fri").last()

Facebook´s stock price on the fourth Friday in January 2015 was ... 77.83!

# Well Done!

-------------------